### DISCUSSION TITLE
Reloading a project

### DISCUSSION THREAD

**Discussion by georgiosd** [waf/CSharpRepl#314] [General]
Hey @waf 

First of all thanks for making this! It's a great tool.

I forked the project and I was trying to make it possible to reload an imported `.csproj` after modifying it, but I've hit a wall.

Basically I did this by:

A. Disabling cache (just in case)
B. Modifying the builder to build to a random folder
C. Clearing previous non-framework references
D. Adding the reference to the built .dll

While this works, I still get the old code running. I checked this by masking a class that returns a value and incrementing this value in the source, between imports.

Any ideas why this may be? I have double checked that the `Document` is created with the right set of references.

**Comment by waf** [comment]
My guess is that the Roslyn workspace is correctly updated, but the Roslyn scripting engine is not. You may need to clear the scriptstate in the scriptrunner. 

This is because the Roslyn scripting libraries are mostly separate from the Roslyn workspace libraries, so CSharpRepl needs to keep them in sync. There's a bit of documentation about that here: https://github.com/waf/CSharpRepl/blob/main/ARCHITECTURE.md#roslyn-libraries

**Reply by georgiosd** [reply]
Hey, thanks for responding. I ended up going to the roslyn discord and discussing this, which was very illuminating.

I mentioned this repro as the desirable outcome:

![image](https://github.com/waf/CSharpRepl/assets/1672054/07ea25cc-dd34-4f98-903f-e3ca34297e8d)

It was then explain to me that this works because of "shadowing", not because of the first `Foo` declaration being replaced by the second. However, any references to the first `Foo` class created before the second one was created would remain tied to the first definition.

Basically, the only way to replace an assembly would be to use `AssemblyLoadContext`s which don't seem to be supported by Roslyn.

In the end they advised me to work with the hot reload feature in VS2022 preview, which actually can do the above.

**Reply by georgiosd** [reply]
@waf looks like someone figured it out: https://github.com/dotnet/roslyn/issues/41722#issuecomment-1784212808
